tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -1114,49 +1114,75 @@ LogicalResult ConvertTFLMaxPool2DOp::matchAndRewrite(
 
 LogicalResult ConvertTFLConv2DOp::matchAndRewrite(
     Operation* op, PatternRewriter& rewriter) const {
-  auto tfl_conv2d_op = cast<TFL::Conv2DOp>(op);
+  Location loc = op->getLoc();
+  auto tfl_conv_op = cast<TFL::Conv2DOp>(op);
 
-  RankedTensorType input_type =
-      tfl_conv2d_op.input().getType().dyn_cast<RankedTensorType>();
-  RankedTensorType filter_type =
-      tfl_conv2d_op.filter().getType().dyn_cast<RankedTensorType>();
-  RankedTensorType output_type =
-      tfl_conv2d_op.getResult().getType().dyn_cast<RankedTensorType>();
-  // Not a ranked tensor output
-  if (!input_type) return failure();
-  if (!output_type) return failure();
-  if (!filter_type) return failure();
+  Value input = tfl_conv_op.input();
+  Value filter = tfl_conv_op.filter();
+  Value bias = tfl_conv_op.bias();
+  Value result = tfl_conv_op.getResult();
+
+  RankedTensorType input_type = input.getType().dyn_cast<RankedTensorType>();
+  RankedTensorType filter_type = filter.getType().dyn_cast<RankedTensorType>();
+  RankedTensorType bias_type = bias.getType().dyn_cast<RankedTensorType>();
+  RankedTensorType output_type = result.getType().dyn_cast<RankedTensorType>();
+
+  // All convolution types should be known rank.
+  if (!input_type || !output_type || !filter_type || !bias_type)
+    return failure();
 
   bool input_is_qtype =
       input_type.getElementType().isa<mlir::quant::QuantizedType>();
   bool filter_is_qtype =
       filter_type.getElementType().isa<mlir::quant::QuantizedType>();
+  bool bias_is_qtype =
+      bias_type.getElementType().isa<mlir::quant::QuantizedType>();
   bool output_is_qtype =
       output_type.getElementType().isa<mlir::quant::QuantizedType>();
 
-  if ((input_is_qtype != filter_is_qtype) ||
-      (input_is_qtype != output_is_qtype)) {
-    return op->emitOpError(
-        "ConvertTFLConv2DOp: input/filter/output tensor should "
-        "be all quantized or all floating-point.");
+  bool all_quantized = input_is_qtype && filter_is_qtype && bias_is_qtype;
+
+  // Find a non-quantized element type if it exists.
+  Type compute_type = input_type.getElementType();
+  if (compute_type.isa<mlir::quant::QuantizedType>())
+    compute_type = filter_type.getElementType();
+  if (compute_type.isa<mlir::quant::QuantizedType>())
+    compute_type = bias_type.getElementType();
+
+  // Make sure the input/filter are either all quantized, or dequantize so that
+  // they are all non-quantized types.
+  if (!all_quantized && input_is_qtype) {
+    Type dequantizedTy =
+        RankedTensorType::get(input_type.getShape(), compute_type);
+    input = rewriter.create<TFL::DequantizeOp>(loc, dequantizedTy, input);
```

Comment:
I'm curious about this - why is it preferable that the insertions of TFL::dequantize a TOSA legalization option, as opposed to some kind of TFL canonicalization step before just performing legalization to TOSA ? 

In addition to looking a little awkward in terms of generating source dialect nodes during legalization, it adds potentially long term maintainability concerns if there are changes in TFL which lead to having to update dependencies elsewhere - here in TOSA legalizations - that need fixing. 

A pass that handles dequantization of mixed precision content on the source dialect side could then be invoked while processing that dialect through the TOSA legalization pipeline.

Reply:
I will try to summarize details as best I can:

1. TFLite actually supports mixed quantization - specifically using quantized weights with floating point computation. This allows weight compression (4x saving on these constants, saved to file) and still improves performance - values are dequantized during the matmul, which can be more computationally efficient than a regular fp matmul.

2. It would not be appropriate for TFLite to canonicalize due to it supporting mixed quantization support, otherwise TFLite would not longer be able to support its mixed precision operations.

3. We insert the TFL::Dequantize operation so that we can reuse its lowering. Both dialect conversion and greedy pass rewriter support recursive patterns. This allows patterns to be reuse and avoids code duplication. It is a very common trait for XLA's lowering, along with core lowering in MLIR.

4. Considering TOSA is in flux and TFLite is not technically standardized, there will always be maintainability concerns. In this case, dequantization is a straightforward operation and quantized/floating point dense operations are also well understood. It is possible that TOSA will support some form of mixed quantized operation however we should not write passes for features that do not exist.

Reply:
So, what I had in mind was an optional pass sitting within TFLite that does --dequantize-mixed-precision-tensor-ops or similar, that feeds any lowering path where such a mixed precision op is simply decomposed using prefixed dequantizes, as is done here. It could even do something trivially in its own tablegen file like: 

def DequantizeConv2DWt : Pat<(TFL_Conv2DOp $input, IsQuantType:$filter, $bias, $dilation, $af, $padding, $stride), 
                                                       (TFL_Conv2DOp $input, (TFL_DequantizeOp $filter), $bias, $dilation, $af, $padding, $stride)>; 

Additionally: 
* I'm not a TableGen expert, probably a single function could implement this, with Conv2D, DW_Conv2D doing the same dequantize insertion on a common TFL_Conv pattern. This is currently duplicated. 
* Enables this to be used by any TFLite consumer that desires to express same-precision variants, e.g. someone has a TFL implementation where the kernels are not mixed precision. 
* It also enables isolation and independent testing of the pattern application. 

Doing this addresses 1) and 2). The code reuse pattern in 3) is rather clever in itself and is not a problem . Item 4) is a concern; as both the source and destination dialects may evolve - it suggests not combining precision changes that may have broader applicability and legalizations to a single dialect, in one function.

Reply:
I have a separate CL that Uses a TFLite pass instead, then includes it as a separate pass in the tfl-to-tosa pipeline. It should handle the decomposition separately. I decided to go the non-tablegen method and instead using a templated rewrite pattern.

Reply:
That sounds great,  sorry for the delay in looking at this!